File: Tree/binary tree paths/PATHS.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int, currentSum: int = 0) -> bool:
        if root is None:
            return False
        currentSum += root.val
        if root.left is None and root.right is None:  # leaf node
            return currentSum == targetSum
        return (self.hasPathSum(root.left, targetSum, currentSum) or
                self.hasPathSum(root.right, targetSum, currentSum)) 

File: Tree/binary tree paths/PATHS.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void ans(vector<string>& vec, string S,TreeNode* root )
    {
        if(root==NULL)
        {
            return;
        }
        if(root->left==NULL && root->right==NULL)
        {
            //leaf
            S= S+ to_string(root->val);
            vec.push_back(S);
            return;
        }
        S+=to_string(root->val) + "->";
        ans(vec,S,root->left);
        ans(vec,S,root->right);
    }
    vector<string> binaryTreePaths(TreeNode* root) 
    {
        vector<string> vec;
        ans(vec, "", root);
        return vec;
    }
};